docs: add compression circuit outline (#4599)
AztecProtocol/aztec-packages: yellow-paper/docs/rollup-circuits/index.md

## Overview

Together with the [validating light node](../l1-smart-contracts/index.md), the rollup circuits must ensure that incoming blocks are valid, that state is progressed correctly, and that anyone can rebuild the state.

To support this, we construct a single proof for the entire block, which is then verified by the validating light node.
This single proof consist of three main components:

> It has **two** sub-trees for transactions, and **one** tree for incoming messages. 

Reviewer: ```suggestion
It has **two** sub-trees for transactions, and **one** tree for L1->L2 messages. 
```
"Incoming" is relative, and always confuses me :)

Together with the [validating light node](../l1-smart-contracts/index.md) the rollup circuits must ensure that incoming blocks are valid, that state is progressed correctly and that anyone can rebuild the state.

To support this, we construct a single proof for the entire block, which is then verified by the validating light node. 

> This single proof is constructed by recursively merging proofs together in two tree subtrees, and then combine these use those trees at the root . 

Reviewer 2: ```suggestion
This single proof is constructed by recursively merging proofs together in two tree subtrees, and then at last combine the final 2 subtrees in a final root circuit.
```

Author: I will update this. Take-away is that we build two trees that we are then using both outputs of in the root, so the "root" is kinda merging two proofs together, but the two proofs are not just the transaction ones (2 merges)

Also. Would like to remove the "rollup" from the circuit name if we can find something better as it seems misleading if it will end up as a validium depending on DA concerns.

This structure allows us to keep the workload of each individual proof small, while making it very parallelizable. 
This works very well for the case where we want many actors to be able to participate in the proof generation.

> The tree structure is outlined below, but the general idea is that we have one subtree tree where all the leaves are transactions (kernel proofs) and through $\log(n)$ steps we can then "compress" them down to just a single root proof.

Reviewer 2: ```suggestion
The tree structure is outlined below, but the general idea is that we have one subtree where all the leaves are transactions (kernel proofs) and through $\log(n)$ steps we can then "compress" them down to just a single root proof.
```

And another tree where all the leaves are messages (cross-chain messages) that is also "compressed"  down to just a single proof going into the final root.

Note that we have two different types of "merger" circuit, depending on what they are compressing.

For transaction compression we have:
- The merge rollup
  - Merges two base rollup proofs OR two merge rollup proofs
- The root rollup
  - Merges two merge rollup proofs

And for the message compression we have:

> - The compress circuit

Reviewer 2: I feel like having merge circuit and compress circuit is not a great naming. Maybe something like msg_compress circuit and tx_compress circuits?

Author: Ye will look at finding some better names for it generally, also want to get rid of the "rollup" part as mentioned above.

>   - Merges `N` compress or base compression proofs

Reviewer 2: Based on the diagram I don't see any node called "base" in the messages sub tree so this should be updated.

Author: Will fix as part of finding some better names.

> In the diagram the size of the tree is limited for show, but a larger tree will have more layers of merge rollups proofs. 

Reviewer 2: ```suggestion
In the diagram the size of the tree is limited for demonstration purposes, but a larger tree would have more layers of merge rollups proofs. 
```
